src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProject.cs

```diff
@@ -0,0 +1,1146 @@
+
+        public void AddMetadataReference(string fullPath, MetadataReferenceProperties properties)
+        {
+            if (string.IsNullOrEmpty(fullPath))
+            {
+                throw new ArgumentException($"{nameof(fullPath)} isn't a valid path.", nameof(fullPath));
+            }
+
+            lock (_gate)
+            {
+                if (ContainsMetadataReference(fullPath, properties))
+                {
+                    throw new InvalidOperationException("The metadata reference has already been added to the project.");
+                }
+
+                _allMetadataReferences.MultiAdd(fullPath, properties);
+
+                if (_activeBatchScopes > 0)
+                {
+                    if (!_metadataReferencesRemovedInBatch.Remove((fullPath, properties)))
+                    {
+                        _metadataReferencesAddedInBatch.Add((fullPath, properties));
+                    }
+                }
+                else
+                {
+                    _workspace.ApplyChangeToWorkspace(w =>
+                    {
+                        var projectReference = _workspace.TryCreateConvertedProjectReference(Id, fullPath, properties);
+
+                        if (projectReference != null)
+                        {
+                            w.OnProjectReferenceAdded(Id, projectReference);
+                        }
+                        else
+                        {
+                            w.OnMetadataReferenceAdded(Id, _workspace.CreateMetadataReference(fullPath, properties));
+                        }
+                    });
+                }
+
+            }
+        }
+
+        public bool ContainsMetadataReference(string fullPath, MetadataReferenceProperties properties)
+        {
+            lock (_gate)
+            {
+                return GetPropertiesForMetadataReference(fullPath).Contains(properties);
+            }
+        }
+
+        /// <summary>
+        /// Returns the properties being used for 
+        /// </summary>
+        /// <param name="fullPath"></param>
+        /// <returns></returns>
+        public ImmutableArray<MetadataReferenceProperties> GetPropertiesForMetadataReference(string fullPath)
+        {
+            lock (_gate)
+            {
+                _allMetadataReferences.TryGetValue(fullPath, out var list);
+
+                // Note: AsImmutableOrEmpty accepts null recievers and treats that as an empty array
+                return list.AsImmutableOrEmpty();
+            }
+        }
+
+        public void RemoveMetadataReference(string fullPath, MetadataReferenceProperties properties)
+        {
+            if (string.IsNullOrEmpty(fullPath))
+            {
+                throw new ArgumentException($"{nameof(fullPath)} isn't a valid path.", nameof(fullPath));
+            }
+
+            lock (_gate)
+            {
+                if (!ContainsMetadataReference(fullPath, properties))
```

Comment:
>ContainsMetadataReference [](start = 21, length = 25)

We could avoid taking the lock again by inlining this.

Reply:
But then there's duplicate logic, and see my comment regarding _NoLock earlier. The tl;dr is there's other optimizations here, and I don't want to microoptimize since we have other work here first.